### PR TITLE
Segment Stacking, need to dedupe the results

### PR DESCRIPTION
**Pull request by Gitzman**

### CODE REVIEW

**Comment by Gitzman** [comment]
Am I doing this right??

**Comment by Gitzman** [comment]
Im just going to keep this open.  I think Im doing it correctly.

**Comment by randyzwitch** [comment]
Yes, this is correct. Now anyone who might want to contribute to this can click the following links to bring this into their repo; when someone fixes the parser, then I'll merge it into master.

screen shot 2016-03-18 at 1 31 05 pm

**Comment by Gitzman** [comment]
So I've got the parser working in cases where # of segments >0.  Now 0 segments is failing.  Also, in the segment name and segment id columns, I concatenate the values and separate them with and AND.  Let me know if you have a different style preference.

**Comment by randyzwitch** [comment]
Why not just do a simple space or comma-separator? You can do AND if you want, just seems like excess effort

**Comment by Gitzman** [comment]
Well I already did the AND.  It's also a bit confusing for somebody who does not know about segment stacking to not explicitly state the AND since the segments are joined with AND logic and not OR logic.

I've got QueueOvertime() functioning without errors or warnings for the use cases of:
   0, 1 and 2 segments
   1 day and multi-day requests

What's up with the Travis thing above?  Did my alterations break something else or is this an unrelated error?

I do need to comment my edits, so I will do that before tomorrow morning.  Let me know if I need to do anything else in order to merge to master. 

I can try my same edits on the other query functions to allow segment stacking universally.

Thanks!!

**Comment by randyzwitch** [comment]
Because it's an authenticated API, the Travis credentials only reside in my account. Nothing to worry about, I'll check it before merging

**Comment by Gitzman** [comment]
Hey,

I just pushed my commented version.  As far as I can tell, segment stacking now works in the QueueOvertime function.  

I did not change very much of the code.  I hope it is as easy to add segment stacking to the other Queue Functions.  I'm going to start to test it out later this week.

**Comment by Gitzman** [comment]
Added segment stacking to all Queue() functions except summary which does not accept a segment id function.  I did sacrifice some logic that removes the segment columns from the returned DF when there is no segment, but that seems worthy compared to the addition of segment stacking.  

Let me know if there is anything I can do to help get this pushed to Master then CRAN.  Thanks!

**Comment by randyzwitch** [comment]
Can you write some tests? I'll need to swap the segments out for ones that exist in my report suite, but it'll be less work for me than writing the tests from scratch

**Comment by randyzwitch** [comment]
Your code breaks the following tests:

```R
  ee <- QueueRanked("zwitchdev", 
                         date.from = "2015-04-01", 
                         date.to = "2015-04-20", 
                         metrics = "pageviews", 
                         elements = c("page", "sitesection") 
  )
  
  ff <- QueueRanked("zwitchdev", 
                         date.from = "2015-04-01", 
                         date.to = "2015-04-20", 
                         metrics = "pageviews", 
                         elements = c("page", "sitesection", "browser") 
  )

  gg <- QueueTrended("zwitchdev", 
                           date.from = "2015-04-01", 
                           date.to = "2015-04-20", 
                           metrics = c("pageviews","visits"), 
                           elements = c("sitesection") 
  )
  
  hh <- QueueTrended("zwitchdev", 
                           date.from = "2015-04-01", 
                           date.to = "2015-04-20", 
                           metrics = "pageviews", 
                           elements = c("sitesection", "page") 
  )

  ii <- QueueTrended("zwitchdev", 
                           date.from = "2015-04-01", 
                           date.to = "2015-04-02", 
                           metrics = "pageviews", 
                           elements = c("sitesection", "page","browser") 
  )
```

All have the same ambiguous error; I believe this occurs when the API returns no results (i.e. no rows for data frame to parse):

```R
 Error in data.frame(..., check.names = FALSE) : 
  arguments imply differing number of rows: 0, 1 
```

**Comment by Gitzman** [comment]
I wrote some tests, I just copied what existed and added 2 segments to a lot of them.  Just pushed.  Let me investigate these failed tests :(

**Comment by Gitzman** [comment]
These all work on my account/report suite.  Is it possible that they never return data and that's why the tests are failing?

Is it possible for me to get temporary credentials to your report suite so I can test?  I don't know what else to do.

**Comment by randyzwitch** [comment]
It's definitely due to no rows being returned. When you run those tests from master, a zero row dataframe is returned with the proper columns.

I'd really prefer to not give access to my account; try to figure out what the logical error might be, and if you can't figure it out, then either I'll fix the issue or set you up with credentials.

**Comment by Gitzman** [comment]
Got ya.  I'll figure out how to handle null returns.  I can't imagine I'll need the credential.  Just wondering, what is important about returning null frames instead of an error when there is no data?

**Comment by randyzwitch** [comment]
Because the package itself is just a conduit; no judgement is made about the validity of a result. Even in the case where the result is a dataframe with zero rows, it was a successful API call and the API returned a fully-formed JSON response. What happens is that there are JSON keys, the values being empty. 

So this isn't an error (which I associate with a programmer syntax error), it's what the API says is the answer.

**Comment by Gitzman** [comment]
Trying to see what I am supposed to return just did the following:

git checkout master
R
>library("devtools")
>load_all(getwd()) (this is how I've been calling my local version of RSiteCat)
>SCAuth()
>QueueTrended(query to retrieve null set)

and I got the same error that I get in my version of the code. I'm fairly certain I am using the master library, because I also tried to run a 2 segment query and it failed. 

Exploring the problem a bit, I found that if I return (formatted.df) from a null after all of the parsing logic in ParseTrended, but before the segment piece at the bottom, it returns a 0x0 dataframe.

Anyhow, if the query doesn't return anything in these functions, what should I return, an empty dataframe with the standard query columns, a 0x0 df or something else?

Let me know.  Thanks for  your help!

File: tests/testthat/test-queuefallout.R [hunk 1]
```diff
@@ -1,24 +1,49 @@
...
+                     element="page",
                      c("http://randyzwitch.com/r-amazon-ec2", "http://randyzwitch.com/julia-import-data")
   )
-  
+
   #Validate returned value is a data.frame
   expect_is(aa, "data.frame")
-  
+
 })
```
Comment (randyzwitch) [inline tests/testthat/test-queuefallout.R]: You have extra `})` values in each function call in the QueueFallout file. There should just be one at the end representing the closing of the `{` bracket in the `test_that` function

Comment (Gitzman) [inline tests/testthat/test-queuefallout.R]: Whoops.  Fixed.

File: tests/testthat/test-queuefallout.R [hunk 2]
```diff
@@ -11,12 +11,11 @@ test_that("Validate QueueFallout using legacy credentials", {
                      metric="pageviews",
                      element="page",
                      c("http://randyzwitch.com/r-amazon-ec2", "http://randyzwitch.com/julia-import-data")
```
Comment (randyzwitch) [inline tests/testthat/test-queuefallout.R]: You are missing the closing `)` on line 13,  and missing commas between the arguments on lines 25 and 38. Please check the validity of your code using RStudio, which will highlight any invalid syntax.

Comment (Gitzman) [inline tests/testthat/test-queuefallout.R]: Apologies, I was not using a good txt editor and not actually running the tests so I couldn't see the areas.  Using RStudio, I corrected the red x's in the Test-QueueFallout file and I checked the other test-queues for red x's and didn't see any.  Pushing update now.

**Comment by randyzwitch** [comment]
Ok, this is starting to feel like hazing, but your logic for when to use `unbox` on segments is incorrect. 

```R
cc <-QueueFallout("zwitchdev",
                  "2014-12-01",
                  "2015-12-31",
                  metrics="pageviews",
                  element="page",
                  checkpoints=c("http://randyzwitch.com/r-amazon-ec2", "http://randyzwitch.com/julia-import-data"),
                  segment.id=c("5433e4e6e4b02df70be4ac63","54adfe3de4b02df70be5ea08"))



 Error: Tried to unbox a vector of length 2 In addition: Warning message:
In if (segment.id != "") { :
  the condition has length > 1 and only the first element will be used
```

**Comment by Gitzman** [comment]
Lol, I feel like some hazing is in order as this is my first Open Source contribution, I just hope that this is still easier than you adding the segment stacking feature yourself!  That does not look like my code as my queuefallout code has:

  #If segment is null, apply the standard segment unbox function
    if(as.list(segment.id)[1]==''){
    report.description$reportDescription$segment_id <- unbox(segment.id)
      }
  #If segment is not null, treat it like a list of metrics.
    else{
    report.description$reportDescription$segments <- data.frame( id = segment.id)

    }

Nowhere in my copy of QueueFallout or in ParseFallout do I see 

if (segment.id != "")

How could this be possible, as I know that test is from my version of tests?

**Comment by randyzwitch** [comment]
To be honest, it's probably at the point where I should merge this in, then find the bug. You can't test against my report suite, and I don't want to give out the credentials. At worst, I can just back the change out of the master branch, no one should be using that anyway.

Thanks for all your work on this, even one more contributor here makes a huge difference on such a small project.

**Comment by randyzwitch** [comment]
It's not your changed that caused the error, it just highlights them. The parser is broken somehow, and I'll need to figure out why, but ultimately #129 can be closed. It just might take a while to get to CRAN

**Pull request merged**